aiomysql/connection.py

```diff
@@ -441,39 +433,50 @@ def set_charset(self, charset):
         """Sets the character set for the current connection"""
         # Make sure charset is supported.
         encoding = charset_by_name(charset).encoding
-        yield from self._execute_command(COMMAND.COM_QUERY, "SET NAMES %s"
-                                         % self.escape(charset))
-        yield from self._read_packet()
-        self._charset = charset
-        self._encoding = encoding
+        yield from self._protocol.execute_command(COMMAND.COM_QUERY,
+                                                  "SET NAMES %s" %
+                                                  self.escape(charset))
+        yield from self._protocol._read_packet()
+        self._protocol._charset = charset
+        self._protocol._encoding = encoding
 
     @asyncio.coroutine
     def _connect(self):
         # TODO: Set close callback
         # raise OperationalError(2006,
         # "MySQL server has gone away (%r)" % (e,))
         try:
+            def proto_lambda():
+                return MySQLProtocol(
+                    host=self._host, port=self._port, user=self._user,
+                    password=self._password,
+                    charset=self._charset, auth_plugin=self._auth_plugin,
+                    use_unicode=self.use_unicode, conv=decoders,
+                    client_flag=self.client_flag, db=self.db,
+                    ssl=self._ssl_context, loop=self.loop
+                )
+
             if self._unix_socket and self._host in ('localhost', '127.0.0.1'):
-                self._reader, self._writer = yield from \
-                    asyncio.open_unix_connection(self._unix_socket,
-                                                 loop=self._loop)
-                self.host_info = "Localhost via UNIX socket: " + \
+                self._socket, self._protocol = yield from \
+                    self.loop.create_unix_connection(proto_lambda,
+                                                     self._unix_socket)
+                self.host_info = "Localhost via UNIX socket: " +\
                                  self._unix_socket
             else:
-                self._reader, self._writer = yield from \
-                    asyncio.open_connection(self._host, self._port,
-                                            loop=self._loop)
+                self._socket, self._protocol = yield from \
+                    self.loop.create_connection(proto_lambda,
```

Comment:
Do we still need `self.loop.create_connection` here I think we can accomplish ssl without using protocol APIs and `MySQLProtocol`. (Separate class for those functions may be good idea, but that should be handled in separate PR after SSL).

Basically my goal is to have minimum not related to ssl directly changes as possible, so code will be easy maintain and review.

Reply:
It should agreed, but I also have no where near enough time to pull out ssl test that using the old code and then PR using protocols.